### ISSUE TITLE
Feature request: hook for running script each time the Guake window become visible

### ISSUE THREAD

**Issue by sm8ps** [Guake/guake#793]
Guake is running well here on Ubuntu 16.04 except for little thing: I use an international keyboard layout in the desktop environment but for the terminal I want to use the US keyboard. I have made the layouts available and can switch between them by a key combination.
However, I have to do this anew each time Guake becomes visible. For some reason the layout is always reset to the standard one. That becomes tiresome quite soon. So I was wondering if there was a way to issue a script for each tab every time that Guake becomes visible.
I have tried the startup script option but of course that script is only run once and does not help much in my case. Also, I toyed around with the idea of running 'guake -s $TAB -e' for each tab but that executes the command directly in the shell which is quite intrusive. Toggling the visibility ten times would clutter each tab with ten identical calls to a script.
Maybe even simpler: would it be possible to select the keyboard layout in the settings?
Cheers!

**Comment by Stibbons** [comment]
If you feel motivated, you can add this feature in settings panel :) should not be that hard, just finding the right API from gtk to do so in Python :)

**Comment by sm8ps** [comment]
Thanks for the encouragement but I most likely would not be able to do that because I am not much of a programmer and do not have any over-sight in the source code of Guake.
Furthermore, I have come to suspect that an additional hook would not solve my problem that easily. I think the problem itself ought to be addressed for which I shall create an error report of its own.

**Comment by mouseratti** [comment]
@sm8ps I think I can implement this

**Comment by Stibbons** [comment]
+1

**Comment by sm8ps** [comment]
That sounds great, @mouseratti! Let me know if I can help testing.

**Comment by mouseratti** [comment]
@Stibbons can you tell me which libs should I install to open **.glade** files with glade3?

I'm trying to do it, but getting such error message:

`(glade:21198): GladeUI-WARNING **: Couldnt recognize GtkBuilder xml, skipping guake.git/data/prefs.glade
`

Google says, that glade3 must work with old **.glade** files, but I can not do that..

**Comment by Stibbons** [comment]
I use glade2 editor

-- 
-----
Gaetan

**Comment by mouseratti** [comment]
@sm8ps what command you'd like to exec on show terminal window. I've written some code and want to test it now

**Comment by sm8ps** [comment]
Thanks for your initiative, @mouseratti! I would like to have the following command run:
```
gsettings set org.gnome.desktop.input-sources current 1
```
This switches the keyboard layout on Ubuntu 16.04 with Unity, cf. some kind of [reference](http://askubuntu.com/questions/604462/how-to-set-languages-shortcuts-from-config-files-or-command-line) on askubuntu.com.
If I may add a wish, it would be nice if the command did not run in the terminal itself, i.e. by launching it on the command line but in the background. For otherwise it would clutter the terminal window everytime it is opened and it probably would not work if the command line were not blank but contained an unfinished entry.
Looking forward to your code!

**Comment by sm8ps** [comment]
Thanks for your initiative, @mouseratti! I would like to have the following command run:
```
gsettings set org.gnome.desktop.input-sources current 1
```
This switches the keyboard layout on Ubuntu 16.04 with Unity, cf. some kind of [reference](http://askubuntu.com/questions/604462/how-to-set-languages-shortcuts-from-config-files-or-command-line) on askubuntu.com.
If I may add a wish, it would be nice if the command did not run in the terminal itself, i.e. by launching it on the command line but in the background. For otherwise it would clutter the terminal window everytime it is opened and it probably would not work if the command line were not blank but contained an unfinished entry.
Looking forward to your code!

**Comment by mouseratti** [comment]
@sm8ps can you test this feature from my branch? I have added new tab into settings, "Hooks". You can set hook on "show" main window event and check it execution.

**Comment by mouseratti** [comment]
#867 is almost finished

**Comment by sm8ps** [comment]
Thanks a lot! I finally managed to compile and install. However, I do not see any tab named "Hooks" (cf. attached screen shot). Could it be because I am using a localized version? I did purge the previous version from the Ubuntu repositories. What can I do?

**Comment by sm8ps** [comment]
Screenshot attached.
![guake-einstellungen_001](https://cloud.githubusercontent.com/assets/3989424/21039717/3c3911fc-bde0-11e6-916c-d43512953370.png)

**Comment by mouseratti** [comment]
@sm8ps which version are you running? Have you started guake from my branch issue-793, from commit **c77b579** or you just updated guake from your operating system packages repository?

**Comment by sm8ps** [comment]
@mouseratti: Forgot to chose the right branch from your Git repo. Sorry, I don't have much experience with software development. Will try again tonight or -morrow and get back.

**Comment by sm8ps** [comment]
OK, got it running. Thanks for your patience! 

The hooked file is executed, indeed. However it does not totally work as expected by itself. Sometimes the keyboard layout is switched and sometimes not. After some tinkering, I found out that adding "sleep 0.1" at the beginning of the script makes it run in a stable fashion. So with this additional tweak my request is fulfilled even though it is not quite Xmas yet. :) Thanks a lot!

I can only guess why the time-out is necessary but I wonder if it would not be a good idea to add such a tiny time-out to the hook itself (I propose 0.2 to be on the safe side) so as to prevent issues that would be hard to track for the end user. What do you think?

**Comment by mouseratti** [comment]
@sm8ps can you show a text of your hook script?

**Comment by sm8ps** [comment]
```stefan@sp1:~$ ll guakerc.sh 
-rwxrwxr-x 1 stefan stefan 238 Dez 11 15:14 guakerc.sh*
stefan@sp1:~$ cat guakerc.sh 
#!/bin/bash
sleep 0.1
echo
echo 'Guake hook'
echo 'was:' $(gsettings get org.gnome.desktop.input-sources current)
gsettings set org.gnome.desktop.input-sources current 1
echo 'is:' $(gsettings get org.gnome.desktop.input-sources current)```
The echoes were put in for testing when launching Guake from a terminal.

**Comment by mouseratti** [comment]
I disabled timeout in your script. Here is my result:
```DEBUG:guake.guake_app:hook on event show has been executed
DEBUG:guake.guake_app:Received window state event: <flags GDK_WINDOW_STATE_WITHDRAWN | GDK_WINDOW_STATE_STICKY | GDK_WINDOW_STATE_ABOVE of type GdkWindowState>
DEBUG:guake.guake_app:Received window state event: <flags GDK_WINDOW_STATE_STICKY | GDK_WINDOW_STATE_ABOVE of type GdkWindowState>
Guake hook
org.gnome.desktop.input-sources was: uint32 2
DEBUG:guake.guake_app:Received window state event: <flags GDK_WINDOW_STATE_ABOVE of type GdkWindowState>
DEBUG:guake.guake_app:Received window state event: <flags GDK_WINDOW_STATE_STICKY | GDK_WINDOW_STATE_ABOVE of type GdkWindowState>
DEBUG:guake.guake_app:Received window state event: <flags GDK_WINDOW_STATE_ABOVE of type GdkWindowState>
DEBUG:guake.guake_app:Received window state event: <flags GDK_WINDOW_STATE_STICKY | GDK_WINDOW_STATE_ABOVE of type GdkWindowState>
org.gnome.desktop.input-sources is: uint32 1
```
It looks working for me without any timeouts. Can you show your log output?

**Comment by sm8ps** [comment]
I did not find any switch for activating debugging messages but here is the output when launching Guake from a terminal. It shows the spurious behavior.
```
INFO:guake.guake_app:Logging configuration complete
/usr/local/lib/python2.7/dist-packages/guake/guake_app.py:866: GtkWarning: IA__gtk_window_set_type_hint: assertion '!gtk_widget_get_mapped (GTK_WIDGET (window))' failed
  self.window.set_type_hint(gtk.gdk.WINDOW_TYPE_HINT_DOCK)
/usr/local/lib/python2.7/dist-packages/guake/guake_app.py:867: GtkWarning: IA__gtk_window_set_type_hint: assertion '!gtk_widget_get_mapped (GTK_WIDGET (window))' failed
  self.window.set_type_hint(gtk.gdk.WINDOW_TYPE_HINT_NORMAL)

Guake hook
was: uint32 1
is: uint32 0

Guake hook
was: uint32 1
is: uint32 1

Guake hook
was: uint32 1
is: uint32 1

Guake hook
was: uint32 1
is: uint32 0

Guake hook
was: uint32 1
is: uint32 1
```
I tried opening the Guake window either by my keyboard shortcut (F4) as well as by clicking the "show"-entry in the Guake menu with identical behavior.

**Comment by mouseratti** [comment]
Looks weird.. Is it full output? Script echo has more than one entry in your log. Have you pushed **F4** multiple times? Can you make this diagnostics:

1. Start guake from another terminal (let's call it **T1**)
2. Press F4 eight times(4 show-hide cycles)
3. Press Ctrl-C inside terminal **T1**
1. Show us  full log output of this session

**Comment by mouseratti** [comment]
Or maybe we could connect by teamviewer, or skype or another tool with ability to share screen and make an online diagnostics?

**Comment by sm8ps** [comment]
Yes, I opened and closed the Guake window multiple times to show the effect. As I said the behavior is no stable, so I had to try a couple of times. Can you try it yourself by adding a different keyboard layout to your desktop environment?
The above output is basically what you request, except that I had started Guake in the background and that I cut the first 22 repeated messages:
```
(guake:520): libglade-WARNING **: unknown attribute `swapped' for <signal>.
```
It seems to be a timing issue between the hook in Guake and whatever is controlling the keyboard layout for new windows; whichever gets to it last wins.
I do not have Skype installed nor do I have any screen sharing available to the outside and, frankly, I am not eager to get it all set up. However, if you feel it really helpful then I can do so (not Skype, though).

**Comment by mouseratti** [comment]
> It seems to be a timing issue between the hook in Guake and whatever is controlling the keyboard layout for new windows; whichever gets to it last wins.
I think so too. It's not a problem of hooks implementation and I think we can close this ticket as resolved.
I can add a link on it to the field hint in the settings. It could help for other users. Are you agree?

**Comment by sm8ps** [comment]
I agree. A hint in the settings would be extremely helpful.
Thank you very much for all your efforts!

**Comment by Stibbons** [comment]
I have not followed everything, if you prepare a PR I can review it

**Comment by sm8ps** [comment]
@mouseratti: Can you please treat @stibbons question? I do not have enough overview of the whole thing, sorry!

**Comment by mouseratti** [comment]
I'll make pull request on week-end

**Comment by mouseratti** [comment]
@Stibbons we can close it as done

**Comment by Stibbons** [comment]
Great :) You know you can put "Fix #793" in the commit message to have github automatically close this ticket on merge :) Magic github !
